railgun/controllers/configuration/secret_utils.go

```diff
@@ -0,0 +1,166 @@
+	// store the ingress record
+	if err := storeRuntimeObject(ctx, c, secret, obj, nsn); err != nil {
+		if errors.IsConflict(err) {
+			log.Error(err, "object updated while reconcilation was running, retrying", nsn.Namespace, nsn.Name)
+			return ctrl.Result{Requeue: true}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	log.Info("kong secret configuration successfully patched patched", "namespace", nsn.Namespace, "name", controllers.ConfigSecretName)
+	return ctrl.Result{}, nil
+}
+
+// isRuntimeObjectSame indicates whether a runtime.Object you intend to store in the configuration secret is the same as what's already stored.
+// This can be used to decide whether or not an update needs to be performed on the configuration secret.
+func isRuntimeObjectSame(secret *corev1.Secret, obj runtime.Object, nsn types.NamespacedName) (bool, error) {
+	// marshal to YAML to check contents
+	cfg, err := yaml.Marshal(obj)
+	if err != nil {
+		return false, err
+	}
+
+	// check if there's any existing object
+	key := configsecret.KeyFor(obj, nsn)
+	if foundCFG, ok := secret.Data[key]; ok {
+		if bytes.Equal(foundCFG, cfg) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// storeRuntimeObject stores a runtime.Object in the configuration secret. Callers should re-queue after this completes successfully.
+func storeRuntimeObject(ctx context.Context, c client.Client, secret *corev1.Secret, obj runtime.Object, nsn types.NamespacedName) error {
+	// marshal to YAML for storage
+	cfg, err := yaml.Marshal(obj)
+	if err != nil {
+		return err
+	}
+
+	// patch the secret with the runtime.Object contents
+	key := configsecret.KeyFor(obj, nsn)
+	secret.Data[key] = cfg
+
+	return c.Update(ctx, secret) // TODO: patch here instead of update for perf
+}
+
+// cleanupObj ensures that a deleted ingress resource is no longer present in the kong configuration secret.
+func cleanupObj(ctx context.Context, c client.Client, log logr.Logger, nsn types.NamespacedName, obj client.Object) (ctrl.Result, error) {
+	// TODO need EVENTS here
+	// TODO need more status updates
+	// TODO: (shane) I want to refactor this into several smaller functions
+	// ^ follow up for these items is in: https://github.com/Kong/kubernetes-ingress-controller/issues/1094
+
+	// get the configuration secret namespace
+	secretNamespace := os.Getenv(controllers.CtrlNamespaceEnv)
+	if secretNamespace == "" {
+		return ctrl.Result{}, fmt.Errorf("kong can not be configured because the required %s env var is not present", controllers.CtrlNamespaceEnv)
+	}
+
+	// grab the configuration secret from the API
+	secret := new(corev1.Secret)
+	if err := c.Get(ctx, types.NamespacedName{Namespace: secretNamespace, Name: controllers.ConfigSecretName}, secret); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	key := configsecret.KeyFor(obj, nsn)
+	if _, ok := secret.Data[key]; ok {
+		delete(secret.Data, key)
+		if err := c.Update(ctx, secret); err != nil { // TODO: patch here instead of update
```

Comment:
I'd opt for making a migration from `Update` to `Patch` a blocker for the 2.0 release.

Reply:
Sounds good, https://github.com/Kong/kubernetes-ingress-controller/issues/1104 was created for this purpose :+1: